File: 2226 maximum-candies-allocated-to-k-children/solution.java
class Solution {
    public int maximumCandies(int[] candies, long k) {
        int l = 1,r = 0, ans = 0;
        for(int i:candies) r=Math.max(r,i);
        while(l<=r) {
            int mid = l+(r-l)/2;
            if(pos(candies,mid,k)) {
                l=mid+1;
                ans=mid;
            } else {
                r=mid-1;
            }            
        }
        return ans;
    }
    private boolean pos(int[] nums,int n,long k) {
        long c = 0;
        for(int i:nums) {
            c+=i/n;
        }
        return c>=k;
    }
}

File: 2226 maximum-candies-allocated-to-k-children/solution.go

func maximumCandies(candies []int, k int64) int {
    l, r, ans := 1, 0, 0
    for _, c := range candies {
        if c > r {
            r = c
        }
    }

    for l <= r {
        mid := l + (r-l)/2
        if pos(candies, mid, k) {
            ans = mid
            l = mid + 1
        } else {
            r = mid - 1
        }
    }
    return ans
}

func pos(nums []int, n int, k int64) bool {
    var c int64 = 0
    for _, v := range nums {
        c += int64(v / n)
    }
    return c >= k
}

